daemon/logger/loggerutils/logfile.go

```diff
@@ -468,117 +517,265 @@ func (w *LogFile) readLogsLocked(currentPos logPos, config logger.ReadConfig, wa
 		Watcher:   watcher,
 		Decoder:   dec,
 		Forwarder: fwd,
-	}).Do(currentFile, currentPos)
+	}).Do(ctx, currentFile, currentPos)
+}
+
+type fileOpener interface {
+	ReaderAt(context.Context) (ra sizeReaderAtCloser, err error)
+	Close()
+	Ref() string
+}
+
+// simpleFileOpener just holds a reference to an already open file
+type simpleFileOpener struct {
+	f      *os.File
+	sz     int64
+	closed bool
+}
+
+func (o *simpleFileOpener) ReaderAt(context.Context) (sizeReaderAtCloser, error) {
+	if o.closed {
+		return nil, errors.New("file is closed")
+	}
+
+	if o.sz == 0 {
+		stat, err := o.f.Stat()
+		if err != nil {
+			return nil, errors.Wrap(err, "error stating file")
+		}
+		o.sz = stat.Size()
+	}
+	return &sizeReaderAtWithCloser{io.NewSectionReader(o.f, 0, o.sz), nil}, nil
+}
+
+func (o *simpleFileOpener) Ref() string {
+	return o.f.Name()
+}
+
+func (o *simpleFileOpener) Close() {
+	_ = o.f.Close()
+	o.closed = true
+}
+
+// converter function used by shareTempFileConverter
+func decompress(dst io.WriteSeeker, src io.ReadSeeker) error {
+	if _, err := src.Seek(0, io.SeekStart); err != nil {
+		return err
+	}
+	rc, err := gzip.NewReader(src)
+	if err != nil {
+		return err
+	}
+	_, err = pools.Copy(dst, rc)
+	if err != nil {
+		return err
+	}
+	return rc.Close()
+}
+
+// compressedFileOpener holds a reference to compressed a log file and will
+// lazily open a decompressed version of the file.
+type compressedFileOpener struct {
+	closed bool
+
+	f *os.File
+
+	lf       *LogFile
+	ifBefore time.Time
+}
+
+func (cfo *compressedFileOpener) ReaderAt(ctx context.Context) (_ sizeReaderAtCloser, retErr error) {
+	_, span := tracing.StartSpan(ctx, "logger.Logfile.Compressed.ReaderAt")
+	defer func() {
+		if retErr != nil {
+			span.SetStatus(retErr)
+		}
+		span.End()
+	}()
+
+	span.SetAttributes(attribute.String("file", cfo.f.Name()))
+
+	if cfo.closed {
+		return nil, errors.New("compressed file closed")
+	}
+
+	gzr, err := gzip.NewReader(cfo.f)
+	if err != nil {
+		return nil, err
+	}
+	defer gzr.Close()
+
+	// Extract the last log entry timestramp from the gzip header
+	// Use this to determine if we even need to read this file based on inputs
+	extra := &rotateFileMetadata{}
+	err = json.Unmarshal(gzr.Header.Extra, extra)
+	if err == nil && !extra.LastTime.IsZero() && extra.LastTime.Before(cfo.ifBefore) {
+		span.SetAttributes(attribute.Bool("skip", true))
+		return &sizeReaderAtWithCloser{}, nil
+	}
+	if err == nil {
+		span.SetAttributes(attribute.Stringer("lastLogTime", extra.LastTime))
+	}
+
+	span.AddEvent("Start decompress")
```

Comment:
I considered this but find that the entries in the daemon logs with this are largely not useful.
I could log this as `Trace` level but then we get nothing in the span since trace logging isn't enable in the logrus logger.

Note, I created that tracing hook 😄